File: sdk/viewer/registry/registry.ts
import { ClientResourceRef } from '@aptre/bldr-sdk/resource/client.js'
import { Resource } from '@aptre/bldr-sdk/resource/resource.js'
import {
  ViewerRegistryResourceService,
  ViewerRegistryResourceServiceClient,
} from './registry_srpc.pb.js'
import {
  RegisterViewerResponse,
  ViewerSurface,
  ListViewersResponse,
  WatchViewersResponse,
  ViewerRegistration,
} from './registry.pb.js'

// ViewerRegistry is a resource that provides viewer registration for plugins.
// Plugins register viewers via registerViewer and watch for changes via watchViewers.
export class ViewerRegistry extends Resource {
  // service is the viewer registry resource service
  private service: ViewerRegistryResourceService

  constructor(resourceRef: ClientResourceRef) {
    super(resourceRef)
    this.service = new ViewerRegistryResourceServiceClient(resourceRef.client)
  }

  // registerViewer registers a viewer for an object type.
  public async registerViewer(
    registration: ViewerRegistration,
    abortSignal?: AbortSignal,
  ): Promise<RegisterViewerResponse> {
    return await this.service.RegisterViewer({ registration }, abortSignal)
  }

  // listViewers returns registered viewers for a surface.
  public async listViewers(
    surface: ViewerSurface,
    abortSignal?: AbortSignal,
  ): Promise<ListViewersResponse> {
    return await this.service.ListViewers({ surface }, abortSignal)
  }

  // watchViewers streams viewer registration changes for a surface.
  public watchViewers(
    surface: ViewerSurface,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchViewersResponse> {
    return this.service.WatchViewers({ surface }, abortSignal)
  }
}

File: sdk/viewer/registry/registry.go
package s4wave_viewer_registry

import (
	"context"

	resource_client "github.com/s4wave/spacewave/bldr/resource/client"
)

// ViewerRegistry is the Go SDK wrapper for the ViewerRegistryResourceService.
type ViewerRegistry struct {
	client  *resource_client.Client
	ref     resource_client.ResourceRef
	service SRPCViewerRegistryResourceServiceClient
}

// NewViewerRegistry creates a new ViewerRegistry resource wrapper.
func NewViewerRegistry(client *resource_client.Client, ref resource_client.ResourceRef) (*ViewerRegistry, error) {
	srpcClient, err := ref.GetClient()
	if err != nil {
		return nil, err
	}
	return &ViewerRegistry{
		client:  client,
		ref:     ref,
		service: NewSRPCViewerRegistryResourceServiceClient(srpcClient),
	}, nil
}

// GetResourceRef returns the resource reference.
func (r *ViewerRegistry) GetResourceRef() resource_client.ResourceRef {
	return r.ref
}

// Release releases the resource reference.
func (r *ViewerRegistry) Release() {
	r.ref.Release()
}

// RegisterViewer registers a viewer for an object type and returns the registration resource ID.
func (r *ViewerRegistry) RegisterViewer(ctx context.Context, reg *ViewerRegistration) (uint32, error) {
	resp, err := r.service.RegisterViewer(ctx, &RegisterViewerRequest{Registration: reg})
	if err != nil {
		return 0, err
	}
	return resp.GetResourceId(), nil
}

// ListViewers returns registered viewers for surface.
func (r *ViewerRegistry) ListViewers(ctx context.Context, surface ViewerSurface) ([]*ViewerRegistration, error) {
	resp, err := r.service.ListViewers(ctx, &ListViewersRequest{Surface: surface})
	if err != nil {
		return nil, err
	}
	return resp.GetRegistrations(), nil
}

// WatchViewers streams viewer registration changes for surface.
func (r *ViewerRegistry) WatchViewers(
	ctx context.Context,
	surface ViewerSurface,
) (SRPCViewerRegistryResourceService_WatchViewersClient, error) {
	return r.service.WatchViewers(ctx, &WatchViewersRequest{Surface: surface})
}
